README
-----------------------------------------------------
	EnderUnix Knowlan - 0.1
    Halil Demirezen <halil@enderunix.org
-----------------------------------------------------


Knowlan is ARP protocol based Local Area Network IP and
MAC Adress Extractor. Knowlan uses libpcap and libnet
libraries for to be simple to handle and to have a simple
code for any interestor to deal with the code. Pre-Stable
Versions of Knowlan did not use libpcap and libnet 
libraries. It used pure link layer system calls to get
settings and generate ARP PACKETs for sending and handling.

However, everything is not so simple. For portability,
It has been found more efficient that if we use a well
known and portable library, we can solve portability problem.
So, It is decided on to use libpcap and libnet libraries 
for sending and handling ARP packets and decide on them
what to do.

As you have already dealt with the program, It has two processess.
One, child enters in a infinite loop for recieving ARP
REPLY packets and then printing IP and MAC Addresses as it 
gets packet from the interface. The second process, our parent,
get a number of maximum possible host in the LAN using netmask.
So, then, it starts sending ARP REQUEST packets to the whole
LAN, every ip, not knowing whether the ip address it sends
the ARP REQUEST packet is up or not.

Arter sending packets to whole LAN, it waits an extra 100msecs. 
for any late host to answer ARP REQUEST. So, after 100msecs 
passed, the send_arp_packet(interface) returns and that process
then sends SIGKILL to the child to inform that there will 
be no other machine to reply. So the child quits. The IP and
MAC addresses associated with them are all active machines
on the LAN. Have fun with them!

For installation please READ INSTALL.

 




ChangeLog
-----------------------------------------------------
        EnderUnix Knowlan - 0.1.1
    Halil Demirezen <halil@enderunix.org
-----------------------------------------------------

 Sat, 10-01-2004 01:45:24 - Halil Demirezen <halil@enderunix.org>
	
	Package: knowlan, Version: 0.1.1
	
	* Pipeable output provided.


 Fri, 19-09-2003 14:53:00 - Halil Demirezen <halil@enderunix.org>

	Package: knowlan, Version: 0.1

        *  Initial Release

 


kn_defs.h
/**************************************************************************
 *   kn_defs.h                                                            *
 *                                                                        *
 *   Copyright (C) 2003 Halil Demirezen  <halil@enderunix.org>            *
 *                                                                        *
 *   kn_defs.h includes arppacket structure for better handling And	  *
 *   includes prototypes of the functions used.				  *
 *                                                                        *
 *   This program is free software; you can redistribute it and/or modify *
 *   it under the terms of the GNU General Public License as published by *
 *   the Free Software Foundation; either version 2, or (at your option)  *
 *   any later version.                                                   *
 *                                                                        *
 *   This program is distributed in the hope that it will be useful,      *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
 *   GNU General Public License for more details.                         *
 *                                                                        *
 *   You should have received a copy of the GNU General Public License    *
 *   along with this program; if not, write to the Free Software          *
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.            *
 *                                                                        *
 **************************************************************************/

#define NORMAL_OUTPUT	0
#define QUIET_OUTPUT	1

struct arppacket {
        u_char destmac[6];
        u_char sourcemac[6];
        unsigned short protocoltype;
        unsigned short hardwaretype;
        unsigned short resprotocol;
        u_char halen;
        u_char palen;
        unsigned short opcode;
        u_char sourceadd[6];
        u_char sourceip[4];
        u_char destadd[6];
        u_char destip[4];
};


int arp(char *);
int send_arp_packet(char *);
int fork(void);
void usage(void);
void version(void);
void wout(void);


kn_main.c
/**************************************************************************
 *   kn_main.c                                                            *
 *                                                                        *
 *   Copyright (C) 2003 Halil Demirezen  <halil@enderunix.org>            *
 *                                                                        *
 *   kn_main.c is the starting point of the program. It forks and with    *
 *   child it recieves ARP replies and with parent it sends ARP packets   *
 *   and after sending ARP reqests finishes, it quits the child by 	  *
 *   killing it.							  *
 *  									  *
 *                                                                        *
 *   This program is free software; you can redistribute it and/or modify *
 *   it under the terms of the GNU General Public License as published by *
 *   the Free Software Foundation; either version 2, or (at your option)  *
 *   any later version.                                                   *
 *                                                                        *
 *   This program is distributed in the hope that it will be useful,      *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
 *   GNU General Public License for more details.                         *
 *                                                                        *
 *   You should have received a copy of the GNU General Public License    *
 *   along with this program; if not, write to the Free Software          *
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.            *
 *                                                                        *
 **************************************************************************/


#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <pcap.h>
#include <errno.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netinet/if_ether.h>
#include <signal.h>

#include "kn_defs.h"


void wout(void);

char *optarg;
int fd[2];
int output_mode = NORMAL_OUTPUT;

int main(int argc, char **argv)
{
     int i, c, error = 0;
     char *interface = NULL;
     char errbuf[PCAP_ERRBUF_SIZE];
     pcap_t * descr;
     char *dev;

		/* if we do not specfify an 
		   interface, let's give it 
		   to pcap to determine  an 
		   interface
   		*/

	if(pipe(fd) < 0)
		perror("Pipe error"), exit(-1);
      
	while(!error && (c = getopt(argc, argv, "i:hvq")) != -1){
		switch(c){
			case 'i':
				interface = optarg;
				break;
			case 'q':
				output_mode = QUIET_OUTPUT;
				break;
			case 'h':			
 				usage();
				return 0;	
			case 'v':
				version();
				return 0;
		}
			if(c == 'i' && !interface) return -1;
	 }

	 if(!interface){	
		interface = pcap_lookupdev(errbuf);
		if(interface == NULL){
		   fprintf(stderr, "can't open device\n");
		   return -1;
		}
	 }

	if(getuid() != 0){
	    fprintf(stderr, "You need root privileges to run knowlan\n");
	    return -1;
	}

     	dev = interface;
	if ((descr = pcap_open_live(dev, 1500, 0, 500, errbuf)) == NULL) {
		fprintf(stderr, "pcap_open_live %s failed: %s\n", dev, errbuf);
		exit(1);
	}
		 /* if we are here, we have got 
		    no problem at all. Let's do 
		    our job 
		 */

	if((i = fork()) == 0){
		/* in the child, there is an 
		   infinite loop that waits 
		   for the arp replies and 
		   printing ip addresses of 
		   the arp replies. 
		*/	
		signal(SIGUSR1, exit);
		arp(interface);
	} else {
		signal(SIGUSR2, wout);
		/* here we did some trick,
		   after sending arp request 
		   packets,  we send   kill 
		   signal to the child which 
		   collects arp replies, on a lan, 
		   arp reply must be less than 100 
		   msec, after sending last packet 
		   and 100msec sleep, we kill child. 
		   That is over. 
		*/
		send_arp_packet(interface);
		kill(i, SIGUSR1);
	}
	return 0;
}

void usage()
{
	fprintf(stderr, "Usage: knowlan [options] [interface]\n");
	fprintf(stderr, "options:\n");
	fprintf(stderr, "-i interface\tact on different interface\n");
	fprintf(stderr, "-h\t\tprints this help\n");
	fprintf(stderr, "-v\t\tversion number\n");
	fprintf(stderr, "-q\t\tA raw output for piping and logging\n");
	fprintf(stderr, "Halil Demirezen <halil@enderunix.org>\n");
}

void version()
{
	fprintf(stderr, "0.2\n");
}

void wout(void)
{
	unsigned char buffer[256];
	if(read(fd[0], buffer, sizeof(buffer)) > 0){
		write(1, buffer, strlen(buffer));
	}
}






kn_arp.c
/**************************************************************************
 *   kn_arp.c                                                             *
 *                                                                        *
 *   Copyright (C) 2003 Halil Demirezen  <halil@enderunix.org>            *
 *   									  *
 *   kn_arp.c includes functions that send and recieve ARP packets to     *
 *   extract IP and MAC address infos of the online machines on the LAN.  *
 *									  *
 *								          *
 *   This program is free software; you can redistribute it and/or modify *
 *   it under the terms of the GNU General Public License as published by *
 *   the Free Software Foundation; either version 2, or (at your option)  *
 *   any later version.                                                   *
 *                                                                        *
 *   This program is distributed in the hope that it will be useful,      *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
 *   GNU General Public License for more details.                         *
 *                                                                        *
 *   You should have received a copy of the GNU General Public License    *
 *   along with this program; if not, write to the Free Software          *
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.            *
 *                                                                        *
 **************************************************************************/



#include <stdio.h>
#include <stdlib.h>
#include <pcap.h>
#include <errno.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netinet/if_ether.h>
#include <libnet.h>

#include "kn_defs.h"

#ifdef __BSD__
#define ether_addr_octet octet
#endif

extern int fd[2];
extern int output_mode;

/* we have to get netmask externall using pcap library. 
   here we can get it. It works, problem  
*/

bpf_u_int32 netmask(char *interface)
{
    bpf_u_int32 netmask;
    bpf_u_int32 ipaddr;
    char *dev;
    char errbuf[PCAP_ERRBUF_SIZE];

    dev = interface;
    pcap_lookupnet(dev, &ipaddr, &netmask, errbuf);
    return(netmask);
}


/* this is an infinite look that listenes for arp packets that are
   ARP_REPLY, then it extracts them. and print the proper information
   to the user. As the options vary, the format and the contents of the
   info vary. This procedure is called as a child process. After 
   parent's packet sending process ends, and after a 100 ms sleep (last  
   sleep of parent, parent sends a SIGKILL signal to this child process.
   Because of having a non SIGKILL handler, child exits. 
*/

int arp(char *interface)
{
	char *dev; 
	char errbuf[PCAP_ERRBUF_SIZE];
	pcap_t* descr;
	const u_char *packet;
	struct pcap_pkthdr hdr;     /* pcap.h */
	struct ether_header *eth;   /* net/ethernet.h */
	struct arppacket *arp;
	struct in_addr tmp;
	unsigned long ip;
	unsigned char buffer[256];
	pid_t parent;
	
	parent = getppid();
	dev = interface;

	if ((descr = pcap_open_live(dev, 1500, 0, 500, errbuf)) == NULL) {
		fprintf(stderr, "pcap_open_live %s failed: %s\n", dev, errbuf);
		exit(1);
     	}
		/* infinite loop for 
		   waiting ARP_REPLY packets.
		*/ 
	if(output_mode == NORMAL_OUTPUT){
		memset(buffer, 0x0, 256);
		sprintf(buffer, "----------------------------------------------\n");
		sprintf(buffer + 47, " EnderUNIX knowlan 0.1.1 - LAN Address Extractor\n");
		sprintf(buffer + 96, " Use -h option for a brief help about the tool\n");
		sprintf(buffer + 143, "----------------------------------------------\n");
		sprintf(buffer + 190, "Interface Name: %s\n", interface);
		write(fd[1], buffer, sizeof(buffer));
		kill(parent, SIGUSR2);
		memset(buffer, 0x0, 256);
		sprintf(buffer, "IP ADRESSES\tMAC ADDRESSES\n");
		write(fd[1], buffer, sizeof(buffer));
		kill(parent, SIGUSR2);
	}

        usleep(300);
	do{
		packet = pcap_next(descr,&hdr);
		eth = (struct ether_header *) packet;
		if (hdr.len < 14)
			continue;
		if (eth == NULL)
			continue;
		if(htons(eth->ether_type) == ETHERTYPE_ARP){
			arp = (struct arppacket *) packet;			
			if(htons(arp->opcode) == ARPOP_REPLY){
				memcpy((char *)&ip, arp->sourceip, 4);
				tmp.s_addr = ip;
				memset(buffer, 0x0, 256);
				sprintf(buffer, "%s\t%02X:%02X:%02X:%02X:%02X:%02X\n",inet_ntoa(tmp), arp->sourceadd[0], arp->sourceadd[1], arp->sourceadd[2],
					        arp->sourceadd[3], arp->sourceadd[4], arp->sourceadd[5]);
				write(fd[1], buffer, sizeof(buffer));
				kill(parent, SIGUSR2);
			}  
		}
	}while(1);
	return 0;
}



int send_arp_packet(char *interface)
{
	struct arppacket *arp;
	u_short packet_size = 42;
	u_char errbuf[256];
	unsigned ip, ip_dst,net;
	u_char *packet;
	struct ether_addr *eth;
	struct libnet_link_int *libnet_link_int;
	u_long number, i;
	char *dev;
 
	if(libnet_init_packet(packet_size, &packet) == -1){
		printf("Can't allocate memory for packet\n");
		return -1;
   	}

	dev = interface;
	libnet_init_packet(packet_size, &packet);
	if ((libnet_link_int = libnet_open_link_interface(interface, errbuf)) == NULL) {
		fprintf(stderr, "cannot open link interface %s: %s\n", interface, errbuf);
		return -1;
  	}

	ip = libnet_get_ipaddr(libnet_link_int, interface, errbuf);
	ip = htonl(ip);
	eth = libnet_get_hwaddr(libnet_link_int, interface, errbuf);
	arp = (struct arppacket *)packet;
		/* let's fill the arp 
		   packet for the proper info 
		*/
	arp->protocoltype = htons(ETHERTYPE_ARP);    /* ARP protocol */
	arp->hardwaretype = htons(ARPHRD_ETHER);     /* Hardware ether */
	arp->resprotocol = htons(ETHERTYPE_IP);      /* Next protocol */
	arp->halen = 6;			             /* Hardare address length */
	arp->palen = 4;				     /* protocol address length */
	arp->opcode = htons(ARPOP_REQUEST);	     /* Arp Opcode */
	memset(arp->destmac, 0xff, 6);	       
	memset(arp->destadd, 0x0, 6);
	memcpy(arp->sourcemac, eth->ether_addr_octet, 6);   /* my ethernet mac */
	memcpy(arp->sourceadd, eth->ether_addr_octet, 6);   /* my arp mac */
	memcpy(arp->sourceip, (char *)&ip, 4);	            /* my arp ip address */	   
	net = netmask(interface);	 		    /* let's get mask */
	number = ntohl(~net);        			    /* maximum number of hosts in the lan */
		/* here send all possible ip 
		   addresses whether they are on 
		   or not. We will decide on this
		   depending on whether they reply 
		   arp back or not 
		*/
	for(i = 1; i < number; i++) {
		ip_dst = (ip & net) | htonl(i);
		memcpy(arp->destip, (char *)&ip_dst, 4);
		if(libnet_write_link_layer(libnet_link_int, interface, (u_char *)packet, packet_size) == -1){
			printf("libnet_write_link_layer failed\n");
			return -1;
		}	
		usleep(100); 
	}
	return 0;
}
